File: libc/openal.c
typedef struct ALCdevice_struct ALCdevice;
typedef struct ALCcontext_struct ALCcontext;

typedef char ALCboolean;
typedef char ALCchar;
typedef char ALCbyte;
typedef unsigned char ALCubyte;
typedef short ALCshort;
typedef unsigned short ALCushort;
typedef int ALCint;
typedef unsigned int ALCuint;
typedef int ALCsizei;
typedef int ALCenum;
typedef float ALCfloat;
typedef double ALCdouble;
typedef void ALCvoid;

ALCdevice* alcOpenDevice(const ALCchar *devicename);
ALCboolean alcCloseDevice(ALCdevice *device);

ALCcontext* alcCreateContext(ALCdevice *device, ALCint *attrlist);
ALCboolean alcMakeContextCurrent(ALCcontext *context);
void alcDestroyContext(ALCcontext *context);

typedef char ALboolean;
typedef char ALchar;
typedef char ALbyte;
typedef unsigned char ALubyte;
typedef short ALshort;
typedef unsigned short ALushort;
typedef int ALint;
typedef unsigned int ALuint;
typedef int ALsizei;
typedef int ALenum;
typedef float ALfloat;
typedef double ALdouble;
typedef void ALvoid;

ALenum alGetError(ALvoid);

ALboolean alIsBuffer(ALuint buffer); 
void alGenBuffers(ALsizei n, ALuint *buffers);
void alDeleteBuffers(ALsizei n, ALuint *buffers);

ALboolean alIsSource(ALuint source);
void alGenSources(ALsizei n, ALuint *sources);
void alDeleteSources(ALsizei n, ALuint *sources);

void alSourcePlay(ALuint source);
void alSourcePause(ALuint source);
void alSourceRewind(ALuint source);
void alSourceStop(ALuint source);

void alSourcef(ALuint source, ALenum param, ALfloat value);
void alSourcei(ALuint source, ALenum param, ALint value);
void alSource3f(ALuint source, ALenum param, ALfloat v1, ALfloat v2, ALfloat v3);

#define AL_GAIN 0x100A
#define AL_PITCH 0x1003
#define AL_LOOPING 0x1007
#define AL_POSITION 0x1004
#define AL_VELOCITY 0x1006

#define AL_TRUE 1
#define AL_FALSE 0

#define AL_BUFFER 0x1009

File: libc/openal.lua
local lib_path
if os.name == 'osx' then 
    lib_path = 'OpenAL.framework/OpenAL'
else
    lib_path = 'System32/OpenAL32'
end

local code, defs = precompile(io.read('./libc/openal.c'))
ffi.cdef(code)

class 'OpenAL' : meta(ffi.load(lib_path))

al = OpenAL()

function OpenAL:setup()
end

function OpenAL:update()
end

function OpenAL:release()
end

local device = al.alcOpenDevice(ffi.NULL)

if device then
    context = al.alcCreateContext(device, ffi.NULL)

    if context then
        al.alcMakeContextCurrent(context)

        local id_ptr = ffi.new('ALuint[1]')
        al.alGenBuffers(1, id_ptr)

        local id = id_ptr[0]
        if al.alIsBuffer(id) then

            local source_ptr = ffi.new('ALuint[1]')
            al.alGenSources(1, source_ptr)

            local source = source_ptr[0]
            if al.alIsSource(source) then
                al.alSourcef(source, al.AL_GAIN, 1)
                al.alSourcef(source, al.AL_PITCH, 1)
                al.alSourcei(source, al.AL_LOOPING, al.AL_FALSE)

                al.alSource3f(source, al.AL_POSITION, 0, 0, 0)
                al.alSource3f(source, al.AL_VELOCITY, 0, 0, 0)

                al.alSourcePlay(source)
                al.alSourceStop(source)
                
                al.alSourcei(source, al.AL_BUFFER, 0)
                
                al.alDeleteSources(1, source_ptr)
            end
            
            al.alDeleteBuffers(1, id_ptr)

        end

        al.alcMakeContextCurrent(ffi.NULL)
        al.alcDestroyContext(context)
    end

    al.alcCloseDevice(device)
end